### ISSUE TITLE
[RFC] Update to StateNotifierProvider's syntax

### ISSUE THREAD

**Issue by rrousselGit** [rrousselGit/river_pod#341]
## Problem

Listening to a `StateNotifierProvider` is a very common source of error for newcomers.

Newcomers tend to write:

```dart
final provider = StateNotifierProvider(...);

Widget build(context, watch) {
  final state = watch(provider).state;

  return Text('$state');
}
```

With this code, the UI will not rebuild when the state of a `StateNotifier` updates.   
While `dartanalyzer` will complain with "state is protected", it appears that many are getting confused by what the warning means and instead proceed to ignore it

## Proposal

The proposal involves the following changes:

1. `StateNotifierProvider` would now take 2 generic parameters (3 with family).

    What previously was:
    ```dart
   class Counter extends StateNotifier<int> {...}
    
   final counterProvider = StateNotifierProvider<Counter>((ref) => Counter());
   ```
   would become:
   ```dart
   class Counter extends StateNotifier<int> {...}
   
   final counterProvider = StateNotifierProvider<Counter, int>((ref) => Counter());
   ```
   where the new generic parameter is the type of the StateNotifier's state.
   
   
2. `watch(stateNotifierProvider)` will now return the StateNotifier's state:

    ```dart
    final counterProvider = StateNotifierProvider<Counter, int>((ref) => Counter());
 
    Widget build(context, watch) {
      int count = watch(counterProvider);     
    }
    ```
    
    This will correctly rebuild the UI when the state updates
    
3.  `watch(stateNotifierProvider.state)` is deprecated
4. A `watch(stateNotifierProvider.notifier)` to obtain the `StateNotifier` , which would not cause the widget to rebuild.

    ```dart
    final counterProvider = StateNotifierProvider<Counter, int>((ref) => Counter());
 
    Widget build(context, watch) {
      Counter counter = watch(counterProvider.notifier);     
    }
    ```

5. (optional) `StateProvider` and `ChangeNotifierProvider` would receive a `watch(provider.notifier)` too, behaving like `StateNotifierProvider` (not causing rebuilds when the state changes)

By doing so:
- the default behaviour of `StateNotifierProvider` will not be confusing to newcomers
- listening to the state of a StateNotifier inside a widget/provider is slightly less verbose (removed a `.state`)
- with 5., this unifies the API for `StateNotifierProvider` vs `StateProvider` vs `ChangeNotifierProvider`, removing further confusion.

The drawback is that creating the provider is slightly more verbose. 

What do you think about this change?

As always, feel free to leave a comment and use 👍 and 👎 to express your opinion.
All feedbacks are welcomed!

**Comment by rrousselGit** [comment]
This change should be fairly simple to automate with a migration tool. 

As such it'd likely come with a script doing the migration for you.

**Comment by arafaysaleem** [comment]
Yes these changes definitely add to the homogeneity of the API for different providers. Being a beginner in both, the state notifier as well riverpod, I can say this is simpler and makes more sense.

**Comment by Marco87Developer** [comment]
I’m in favor of all 5 proposed changes.

**Comment by julienlebren** [comment]
I completely agree with these changes.

I'm a newcomer and even if I didn't make the mistake you quoted, I was quite uncomfortable with the ".state" notation while reading a provider, even if that's just a matter of code beauty.

Things would become clearer with this new proposal. Go!

**Comment by scopendo** [comment]
This seems positive. A call to watch returns the value of the provider.

I may have missed it but I think for consistency, although a breaking change, that the same should be true for a StateProvider so that

```dart
final value = watch(myStateProvider).state;
```

Becomes

```dart
final value = watch(myStateProvider);
```

If the separate RFC for introducing a consumer reference goes ahead, then I'd prefer a different method to watch to obtain the notifier, as I think that will be less confusing.

**Comment by Solido** [comment]
I really welcome the proposition.

When working on large app RiverPod is the tool that should help to read the business modelisation and this work well with
`final serviceProvider = StateNotifierProvider<ServiceImplementation, ServiceData>((ref) => ServiceImplementation());`
It is considered descriptive and not verbose.

Concerning the semantics `Counter counter = *watch*(counterProvider.notifier); ` 
We now have a **watch** that does not rebuild.  I would prefer changing the verb that adding an attribute if possible.

Thanks Remi.

**Comment by rrousselGit** [comment]
> We now have a watch that does not rebuild. I would prefer changing the verb that adding an attribute if possible.

Watching the notifier _does_ cause rebuilds: when the notifier changes

This can happen when the notifier is recreated, such as with "refresh". 

So watch is the correct verb

**Comment by Solido** [comment]
Thanks for clarifying my comprehension of point 4,  All good for me now.

>  _A` watch(stateNotifierProvider.notifier) to obtain the StateNotifier , **which would not cause the widget to rebuild**._

**Comment by thumbert** [comment]
This one is the only problematic one in my view
```
 Counter counter = watch(counterProvider.notifier); 
```
The watch/read dichotomy in provider is clear.  Here you 'watch' but you don't update.  Would be good to use a different verb.

Thanks for your work on this!

**Comment by rrousselGit** [comment]
This was explained previously: https://github.com/rrousselGit/river_pod/issues/341#issuecomment-782700274

It does "watch" something: the notifier
It's just that the notifier rarely changes. But sometimes it does.

**Comment by derolf** [comment]
I was about to open exactly an issue about this.

I exactly fell into that pitfall and can imagine that introducing it to the bigger team we will see hundreds of mistakes cause by this.

Would it be possible to already have a new class that implements the new state/notifier syntax?

**Comment by TimWhiting** [comment]
@rrousselGit
A few notes as I was working on the auto migration of this syntax update (even though this feature is not available yet):

All ChangeNotifiers will not change, because this introduces a new behavior for just listening to changes of the identity of the notifier, so I assume that you will always listen to changes of the actual state of the ChangeNotifier.

All StateProviders would be updated as follows
```dart
// Old
final StateController countNotifier = watch(counterProvider);
final int count = watch(counterProvider).state;

// New
final StateController countNotifier = watch(counterProvider.notifier);
final int count = watch(counterProvider);
```
However, this breaks old behavior, since as far as I understand `watch(counterProvider)` used to trigger a rebuild both when the `StateController` changed or when the state of the `StateController` (in this case an `int`) changed. In the new behavior if you are getting the `StateController` rather than the underlying state you will only be notified if the identity of the notifier changes. Developers might be depending on that rebuild behavior, and an auto-migration tool likely cannot determine what the intention of the developer was.

**Comment by rrousselGit** [comment]
> All ChangeNotifiers will not change

Indeed this is not breaking for ChangeNotifier. There's no need to migrate anything



> However, this breaks old behavior, since as far as I understand `watch(counterProvider)` used to trigger a rebuild both when the `StateController` changed or when the state of the `StateController` (in this case an `int`) changed. In the new behavior if you are getting the `StateController` rather than the underlying state you will only be notified if the identity of the notifier changes. Developers might be depending on that rebuild behavior, and an auto-migration tool likely cannot determine what the intention of the developer was.

`watch(stateProvider)` should still return a `StateController<T>`, so this is not breaking.

The only change is the addition of `watch(stateProvider.notifier)`, which is not a breaking change.

**Comment by venkatd** [comment]
I'm in full support of the changes.

I've been using riverpod for a while, and even though I know the difference between `watch(provider).state` and `watch(provider.state)`, I still make mistakes and have "why isn't this widget rebuilding?" moments.

I'm OK with the drawback of the extra type parameter for the benefit of better defaults.

**Comment by rrousselGit** [comment]
This was implemented as part of version 0.14.0

A migration guide was written, accompanied by a migration tool (thanks to @TimWhiting)
See https://riverpod.dev/docs/migration/0.13.0_to_0.14.0

**Comment by fpabl0** [comment]
@rrousselGit 
This is a good change, but I have a doubt. I think now the API is not consistence as @scopendo said. If `watch(notifierProvider)` listen for just the state, I think `watch(stateProvider)` should listen for the state too and not the `StateController`, at the end `StateController` is also a `StateNotifier` so it is a bit confusing. Why do they have such differences?
